File: scripts/asr/transcribe_file.py
# SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: MIT

import argparse

import os
import riva.client
from riva.client.argparse_utils import add_asr_config_argparse_parameters, add_connection_argparse_parameters


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Streaming transcription of a file via Riva AI Services. Streaming means that audio is sent to a "
        "server in small chunks and transcripts are returned as soon as these transcripts are ready. "
        "You may play transcribed audio simultaneously with transcribing by setting one of parameters "
        "`--play-audio` or `--output-device`.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--input-file", help="A path to a local file to stream.")
    group.add_argument("--list-models", action="store_true", help="List available models.")
    group.add_argument("--list-devices", action="store_true", help="List output devices indices")

    parser.add_argument(
        "--show-intermediate", action="store_true", help="Show intermediate transcripts as they are available."
    )
    parser.add_argument(
        "--output-device",
        type=int,
        default=None,
        help="Output audio device to use for playing audio simultaneously with transcribing. If this parameter is "
        "provided, then you do not have to `--play-audio` option."
    )
    parser.add_argument(
        "--play-audio",
        action="store_true",
        help="Whether to play input audio simultaneously with transcribing. If `--output-device` is not provided, "
        "then the default output audio device will be used.",
    )
    parser.add_argument(
        "--file-streaming-chunk",
        type=int,
        default=1600,
        help="A maximum number of frames in one chunk sent to server.",
    )
    parser.add_argument(
        "--simulate-realtime",
        action='store_true',
        help="Option to simulate realtime transcription. Audio fragments are sent to a server at a pace that mimics "
        "normal speech.",
    )
    parser.add_argument(
        "--print-confidence", action="store_true", help="Whether to print stability and confidence of transcript. If `--word-time-offsets` or `--speaker-diarization` is set, then confidence is not printed."
    )
    parser = add_connection_argparse_parameters(parser)
    parser = add_asr_config_argparse_parameters(parser, max_alternatives=True, profanity_filter=True, word_time_offsets=True)
    args = parser.parse_args()
    if args.play_audio or args.output_device is not None or args.list_devices:
        import riva.client.audio_io
    return args


def main() -> None:
    args = parse_args()
    if args.list_devices:
        riva.client.audio_io.list_output_devices()
        return
    auth = riva.client.Auth(args.ssl_cert, args.use_ssl, args.server, args.metadata)
    asr_service = riva.client.ASRService(auth)

    if args.list_models:
        asr_models = dict()
        config_response = asr_service.stub.GetRivaSpeechRecognitionConfig(riva.client.proto.riva_asr_pb2.RivaSpeechRecognitionConfigRequest())
        for model_config in config_response.model_config:
            if model_config.parameters["type"] == "online":
                language_code = model_config.parameters['language_code']
                model = {"model": [model_config.model_name]}
                if language_code in asr_models:
                    asr_models[language_code].append(model)
                else:
                    asr_models[language_code] = [model]

        print("Available ASR models")
        asr_models = dict(sorted(asr_models.items()))
        print(asr_models)
        return

    if not os.path.isfile(args.input_file):
        print(f"Invalid input file path: {args.input_file}")
        return

    config = riva.client.StreamingRecognitionConfig(
        config=riva.client.RecognitionConfig(
            language_code=args.language_code,
            model=args.model_name,
            max_alternatives=1,
            profanity_filter=args.profanity_filter,
            enable_automatic_punctuation=args.automatic_punctuation,
            verbatim_transcripts=not args.no_verbatim_transcripts,
            enable_word_time_offsets=args.word_time_offsets or args.speaker_diarization,
        ),
        interim_results=True,
    )
    riva.client.add_word_boosting_to_config(config, args.boosted_lm_words, args.boosted_lm_score)
    riva.client.add_speaker_diarization_to_config(config, args.speaker_diarization, args.diarization_max_speakers)
    riva.client.add_endpoint_parameters_to_config(
        config,
        args.start_history,
        args.start_threshold,
        args.stop_history,
        args.stop_history_eou,
        args.stop_threshold,
        args.stop_threshold_eou
    )
    riva.client.add_custom_configuration_to_config(
        config,
        args.custom_configuration
    )
    sound_callback = None
    try:
        if args.play_audio or args.output_device is not None:
            wp = riva.client.get_wav_file_parameters(args.input_file)
            sound_callback = riva.client.audio_io.SoundCallBack(
                args.output_device, wp['sampwidth'], wp['nchannels'], wp['framerate'],
            )
            delay_callback = sound_callback
        else:
            delay_callback = riva.client.sleep_audio_length if args.simulate_realtime else None
        with riva.client.AudioChunkFileIterator(
            args.input_file, args.file_streaming_chunk, delay_callback,
        ) as audio_chunk_iterator:
            riva.client.print_streaming(
                responses=asr_service.streaming_response_generator(
                    audio_chunks=audio_chunk_iterator,
                    streaming_config=config,
                ),
                show_intermediate=args.show_intermediate,
                additional_info="time" if (args.word_time_offsets or args.speaker_diarization) else ("confidence" if args.print_confidence else "no"),
                word_time_offsets=args.word_time_offsets or args.speaker_diarization,
                speaker_diarization=args.speaker_diarization,
            )
    finally:
        if sound_callback is not None and sound_callback.opened:
            sound_callback.close()


if __name__ == "__main__":
    main()

File: riva-ts-client/scripts/asr/transcribe-file.ts


import * as fs from 'fs';
import * as path from 'path';
import * as wavPlayer from 'node-wav-player';
import { program } from 'commander';
import { Auth, ASRService } from '../../src/client';
import { addConnectionArgparseParameters } from '../utils/argparse';
import { addAsrConfigArgparseParameters } from '../utils/asr_argparse';
import { AudioEncoding, StreamingRecognitionConfig, AudioChunk } from '../../src/client/asr/types';
import { getWavFileParameters } from '../../src/client/asr/utils';

class AudioPlayer {
    private tempFile: string;

    constructor() {
        this.tempFile = path.join(process.cwd(), '.temp_audio.wav');
    }

    async play(audioData: Buffer): Promise<void> {
        fs.writeFileSync(this.tempFile, audioData);
        try {
            await wavPlayer.play({ path: this.tempFile });
        } catch (error) {
            console.error('Error playing audio:', error);
        }
    }

    close(): void {
        if (fs.existsSync(this.tempFile)) {
            fs.unlinkSync(this.tempFile);
        }
    }
}

async function* createAudioSource(fileStream: fs.ReadStream): AsyncGenerator<AudioChunk, void, unknown> {
    try {
        for await (const chunk of fileStream) {
            if (chunk instanceof Buffer && chunk.length > 0) {
                yield { audioContent: chunk };
            }
        }
    } finally {
        fileStream.destroy();
    }
}

async function main() {
    program
        .description(
            'Streaming transcription of a file via Riva AI Services. Streaming means that audio is sent to a ' +
            'server in small chunks and transcripts are returned as soon as these transcripts are ready. ' +
            'You may play transcribed audio simultaneously with transcribing by setting --play-audio option.'
        )
        .option('--input-file <file>', 'A path to a local file to stream.')
        .option('--list-models', 'List available models.')
        .option('--show-intermediate', 'Show intermediate transcripts as they are available.')
        .option('--play-audio', 'Whether to play input audio simultaneously with transcribing.')
        .option('--file-streaming-chunk <number>', 'A maximum number of frames in one chunk sent to server.', '1600')
        .option(
            '--simulate-realtime',
            'Option to simulate realtime transcription. Audio fragments are sent to a server at a pace that mimics normal speech.'
        );

    addConnectionArgparseParameters(program);
    addAsrConfigArgparseParameters(program);

    program.parse();

    const options = program.opts();

    if (!options.inputFile && !options.listModels) {
        console.error('Either --input-file or --list-models must be specified');
        process.exit(1);
    }

    const auth = new Auth({
        uri: options.server,
        useSsl: options.useSsl,
        sslCert: options.sslCert,
        metadata: options.metadata?.map(m => {
            const [key, value] = m.split('=');
            return [key, value] as [string, string];
        })
    });

    const asr = new ASRService({
        serverUrl: options.server,
        auth
    });

    if (options.listModels) {
        try {
            const models = await asr.listModels();
            console.log('Available models:');
            for (const model of models) {
                console.log(`  ${model.name}`);
                console.log(`    Languages: ${model.languages.join(', ')}`);
                console.log(`    Sample Rate: ${model.sampleRate}Hz`);
                console.log(`    Streaming: ${model.streaming}`);
                console.log();
            }
            return;
        } catch (error) {
            console.error('Error listing models:', error);
            process.exit(1);
        }
    }

    const inputFile = path.resolve(options.inputFile);
    if (!fs.existsSync(inputFile)) {
        console.error(`Input file ${inputFile} does not exist`);
        process.exit(1);
    }

    const { encoding, sampleRate } = await getWavFileParameters(inputFile);
    if (encoding !== AudioEncoding.LINEAR_PCM) {
        console.error('Only LINEAR_PCM WAV files are supported');
        process.exit(1);
    }

    const fileStream = fs.createReadStream(inputFile);
    let audioPlayer: AudioPlayer | null = null;

    if (options.playAudio) {
        audioPlayer = new AudioPlayer();
    }

    try {
        const config: StreamingRecognitionConfig = {
            config: {
                encoding,
                sampleRateHertz: sampleRate,
                languageCode: options.languageCode || 'en-US',
                maxAlternatives: options.maxAlternatives ? parseInt(options.maxAlternatives) : 1,
                profanityFilter: options.profanityFilter,
                enableWordTimeOffsets: options.wordTimeOffsets,
                enableAutomaticPunctuation: options.automaticPunctuation
            }
        };

        const audioSource = createAudioSource(fileStream);
        const responses = await asr.streamingRecognize(audioSource, config);

        for await (const response of responses) {
            if (response.results.length > 0) {
                const result = response.results[0];
                if (result.alternatives.length > 0) {
                    const transcript = result.alternatives[0].transcript;
                    if (!result.isPartial) {
                        console.log(`Final transcript: ${transcript}`);
                    } else if (options.showIntermediate) {
                        console.log(`Intermediate transcript: ${transcript}`);
                    }
                }
            }

            if (audioPlayer && response.audioContent) {
                const audioBuffer = Buffer.from(response.audioContent);
                await audioPlayer.play(audioBuffer);
            }
        }
    } catch (error) {
        console.error('Error in transcription:', error);
        process.exit(1);
    } finally {
        if (audioPlayer) {
            audioPlayer.close();
        }
    }
}

if (require.main === module) {
    main().catch(console.error);
}
